[Workers] Simplify early dedicated worker termination

Ahmad-S792 opened the pull request:
#### 0987e578a6a8813156f74ebbf01490e4d31aa8f8
<pre>
[Workers] Simplify early dedicated worker termination

[Workers] Simplify early dedicated worker termination
<a href="https://bugs.webkit.org/show_bug.cgi?id=248173">https://bugs.webkit.org/show_bug.cgi?id=248173</a>

Reviewed by Darin Adler.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=180532

When the worker was already asked to terminate by the time script content is loaded,
there is no need to create and start worker thread, we could exit early instead.

* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WorkerMessagingProxy::startWorkerGlobalScope): Add condition to terminate worker early
(WorkerMessagingProxy::setResourcCachingDisabledByWebInspector): Remove worker termination logic since moved above, also add ASSERT and update value of &quot;queuedEarlyTasks&quot;

Canonical link: <a href="https://commits.webkit.org/256949@main">https://commits.webkit.org/256949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4182b20bb3e20657b603c240b9d488302d869b0c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97206 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106724 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166995 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101175 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6756 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35205 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89687 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103412 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102872 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5081 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83832 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32133 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75059 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/494 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/476 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21683 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4788 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44177 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1717 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41002 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/workers/WorkerMessagingProxy.cpp:
```diff
@@ -325,21 +330,17 @@ void WorkerMessagingProxy::setResourceCachingDisabledByWebInspector(bool disable
+        m_workerThread->suspend();
     }
+
+    auto queuedEarlyTasks = WTFMove(m_queuedEarlyTasks);
```
Because we are touching this it should use be changed to `std::exchange(m_queuedEarlyTasks, { })` rather than `WTFMove` since it’s relying on `m_queuedEarlyTasks` becoming empty.

Ahmad-S792 replied:
Done

Ahmad-S792 committed:
[Workers] Simplify early dedicated worker termination

[Workers] Simplify early dedicated worker termination
https://bugs.webkit.org/show_bug.cgi?id=248173

Reviewed by Darin Adler.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=180532

When the worker was already asked to terminate by the time script content is loaded,
there is no need to create and start worker thread, we could exit early instead.

* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WorkerMessagingProxy::startWorkerGlobalScope): Add condition to terminate worker early
(WorkerMessagingProxy::setResourcCachingDisabledByWebInspector): Remove worker termination logic since moved above, also add ASSERT and update value of "queuedEarlyTasks"

Canonical link: https://commits.webkit.org/256949@main
